From b1247696cfd51eaad01c2fdbad57fe09f7f6cf72 Mon Sep 17 00:00:00 2001
From: Thiago Macieira <thiago.macieira@intel.com>
Date: Fri, 15 Aug 2025 15:24:37 -0700
Subject: [PATCH 138/553] QByteArray: match QString::indexOf behavior for large
 negative "from"

Amends b347d487048cf36d609dd31952dbef1813b5b0e5.

Drive-by change the unit test to use QCOMPARE and add lastIndexOf()
tests for good measure.

Change-Id: I474e2d6060bf10179545fffd9ce2caa517114b48
Reviewed-by: Edward Welbourne <edward.welbourne@qt.io>
(cherry picked from commit 48c3c81df09282899d2bf69bb78c9265f8e16ad7)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/text/qbytearrayview.h             |  4 +-
 .../tools/collections/tst_collections.cpp     | 46 +++++++++++++++----
 2 files changed, 39 insertions(+), 11 deletions(-)

diff --git a/src/corelib/text/qbytearrayview.h b/src/corelib/text/qbytearrayview.h
index 4f40fd12097..ea3fc1bcd15 100644
--- a/src/corelib/text/qbytearrayview.h
+++ b/src/corelib/text/qbytearrayview.h
@@ -405,8 +405,10 @@ inline quint16 qChecksum(const char *s, qsizetype len,
 
 qsizetype QtPrivate::findByteArray(QByteArrayView haystack, qsizetype from, char needle) noexcept
 {
+    if (from < -haystack.size()) // from < 0 && abs(from) > haystack.size(), avoiding overflow
+        return -1;
     if (from < 0)
-        from = qMax(from + haystack.size(), qsizetype(0));
+        from = from + haystack.size();
     if (from < haystack.size()) {
         const char *const b = haystack.data();
         if (const auto n = static_cast<const char *>(
diff --git a/tests/auto/corelib/tools/collections/tst_collections.cpp b/tests/auto/corelib/tools/collections/tst_collections.cpp
index cf5a0b692e9..463f600e36d 100644
--- a/tests/auto/corelib/tools/collections/tst_collections.cpp
+++ b/tests/auto/corelib/tools/collections/tst_collections.cpp
@@ -1113,11 +1113,24 @@ void tst_Collections::byteArray()
     QVERIFY (true == hello.contains('e'));
     QVERIFY (hello.contains('e') != false);
 
-    QVERIFY(hello.indexOf('e') == 1);
-    QVERIFY(hello.indexOf('e', -10) == 1);
-    QVERIFY(hello.indexOf('l') == 2);
-    QVERIFY(hello.indexOf('l',2) == 2);
-    QVERIFY(hello.indexOf('l',3) == 3);
+    QCOMPARE(hello.indexOf('e'), 1);
+    QCOMPARE(hello.indexOf('e', -3), -1);   // searches "llo"
+    QCOMPARE(hello.indexOf('e', -4), 1);    // searches "ello"
+    QCOMPARE(hello.indexOf('e', -6), -1);   // overflows size search
+    QCOMPARE(hello.indexOf('l'), 2);
+    QCOMPARE(hello.indexOf('l', 2), 2);
+    QCOMPARE(hello.indexOf('l', 3), 3);
+    QCOMPARE(hello.indexOf('l', 4), -1);
+    QCOMPARE(hello.indexOf('l', 6), -1);
+
+    QCOMPARE(hello.lastIndexOf('e'), 1);
+    QCOMPARE(hello.lastIndexOf('e', -4), 1);    // searches "he"
+    QCOMPARE(hello.lastIndexOf('e', -5), -1);   // searches "h"
+    QCOMPARE(hello.lastIndexOf('e', -6), -1);   // overflows size search
+    QCOMPARE(hello.lastIndexOf('l'), 3);
+    QCOMPARE(hello.lastIndexOf('l', 1), -1);     // searches "he"
+    QCOMPARE(hello.lastIndexOf('l', 2), 2);      // searches "hel"
+    QCOMPARE(hello.lastIndexOf('l', 3), 3);      // searches "hell"
 
     QByteArray empty;
     QCOMPARE(empty.indexOf("x"), -1);
@@ -1860,11 +1873,24 @@ void tst_Collections::qstring()
     QVERIFY (true == hello.contains('e'));
     QVERIFY (hello.contains('e') != false);
 
-    QVERIFY(hello.indexOf('e') == 1);
-    QVERIFY(hello.indexOf('e', -10) == -1);
-    QVERIFY(hello.indexOf('l') == 2);
-    QVERIFY(hello.indexOf('l',2) == 2);
-    QVERIFY(hello.indexOf('l',3) == 3);
+    QCOMPARE(hello.indexOf('e'), 1);
+    QCOMPARE(hello.indexOf('e', -3), -1);   // searches "llo"
+    QCOMPARE(hello.indexOf('e', -4), 1);    // searches "ello"
+    QCOMPARE(hello.indexOf('e', -6), -1);   // overflows size search
+    QCOMPARE(hello.indexOf('l'), 2);
+    QCOMPARE(hello.indexOf('l', 2), 2);
+    QCOMPARE(hello.indexOf('l', 3), 3);
+    QCOMPARE(hello.indexOf('l', 4), -1);
+    QCOMPARE(hello.indexOf('l', 6), -1);
+
+    QCOMPARE(hello.lastIndexOf('e'), 1);
+    QCOMPARE(hello.lastIndexOf('e', -4), 1);    // searches "he"
+    QCOMPARE(hello.lastIndexOf('e', -5), -1);   // searches "h"
+    QCOMPARE(hello.lastIndexOf('e', -6), -1);   // overflows size search
+    QCOMPARE(hello.lastIndexOf('l'), 3);
+    QCOMPARE(hello.lastIndexOf('l', 1), -1);     // searches "he"
+    QCOMPARE(hello.lastIndexOf('l', 2), 2);      // searches "hel"
+    QCOMPARE(hello.lastIndexOf('l', 3), 3);      // searches "hell"
 
     QString empty;
     QCOMPARE(empty.indexOf("x"), -1);
-- 
2.51.2

